### PR TITLE
Add external dependencies to release package

### PR DESCRIPTION
**Pull request by ausfeldt**
## Description

Adding external cosmos dependencies to release package so that we have a single package. Still working on adding tests/validation for CD pipeline.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj [hunk 1]
```diff
@@ -44,15 +44,34 @@
...
+    <PackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="4.5.1" />
+    <PackageReference Include="System.Threading.Tasks.Extensions" Version="4.5.1" />
+    <PackageReference Include="System.ValueTuple" Version="4.5.0" />
   </ItemGroup>
 
-  <ItemGroup Condition=" '$(SignAssembly)' != 'true' ">
-    <PackageReference Include="Microsoft.Azure.Cosmos.Direct.MyGet" Version="[$(DirectVersion)]" />
+  <ItemGroup Condition=" '$(SignAssembly)' == 'true' ">
+    <!--CD will ensure below are present before signed release build-->
+    <None Include="$(MSBuildThisFileDirectory)\Cosmos.CRTCompat.dll" Pack="true" IsAssembly="true" PackagePath="runtimes\win-x64\native" />
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: How will this work with Azure functions or other places that do not support custom dll to be loaded?

Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: This PR was requested by @kirankumarkolli

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: End results is still the same, its only packaging that differs here. 
Issues with native custom DLL's if-any are orthogonal.

**Comment by ausfeldt** [comment]
We decided not to do this.

**Comment by kirankumarkolli** [comment]
Remove native\* folder.

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj [hunk 2]
```diff
@@ -44,15 +44,38 @@
...
-  <ItemGroup Condition=" '$(SignAssembly)' == 'true' ">
-    <PackageReference Include="Microsoft.Azure.Cosmos.Direct" Version="[$(DirectVersion)]" />
-  </ItemGroup>
-
   <ItemGroup Condition=" '$(SignAssembly)' != 'true' ">
     <PackageReference Include="Microsoft.Azure.Cosmos.Direct.MyGet" Version="[$(DirectVersion)]" />
+    <PackageReference Include="Microsoft.Azure.Cosmos.Serialization.HybridRow" Version="1.0.0-preview" />
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(SignAssembly)' == 'true' ">
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: >>     [](start = 54, length = 5)

extra spaces

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj [hunk 3]
```diff
@@ -44,15 +44,38 @@
...
+    <PackageReference Include="Microsoft.VisualStudio.Threading.Analyzers" Version="16.0.102" PrivateAssets="All" />    
   </ItemGroup>
   
-  <ItemGroup Condition=" '$(SignAssembly)' == 'true' ">
-    <PackageReference Include="Microsoft.Azure.Cosmos.Direct" Version="[$(DirectVersion)]" />
-  </ItemGroup>
-
   <ItemGroup Condition=" '$(SignAssembly)' != 'true' ">
     <PackageReference Include="Microsoft.Azure.Cosmos.Direct.MyGet" Version="[$(DirectVersion)]" />
+    <PackageReference Include="Microsoft.Azure.Cosmos.Serialization.HybridRow" Version="1.0.0-preview" />
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: HR is signed always. Can be unconditional.

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj [hunk 4]
```diff
@@ -44,15 +44,38 @@
...
+      <HintPath>$(MSBuildThisFileDirectory)\Microsoft.Azure.Cosmos.Direct.dll</HintPath>
+    </Reference>
+    <!--Direct Dependencies-->
+    <PackageReference Include="System.Configuration.ConfigurationManager" Version="4.5.0" />
+    <PackageReference Include="System.ServiceModel.Primitives" Version="4.5.0" />
+    <!--HybridRow Dependencies-->
+    <PackageReference Include="System.Memory" Version="4.5.1" />
+    <PackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="4.5.1" />
+    <PackageReference Include="System.Threading.Tasks.Extensions" Version="4.5.1" />
+    <PackageReference Include="System.ValueTuple" Version="4.5.0" />
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: Unconditional

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj [hunk 5]
```diff
@@ -44,15 +44,38 @@
...
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(SignAssembly)' == 'true' ">
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: Why conditional. we could generate an unsigned package right?

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj [hunk 6]
```diff
@@ -44,15 +44,38 @@
...
+    <!--CD will ensure below are present before signed release build-->
+    <None Include="$(MSBuildThisFileDirectory)\Cosmos.CRTCompat.dll" Pack="true" IsAssembly="true" PackagePath="runtimes\win-x64\native" />
+    <None Include="$(MSBuildThisFileDirectory)\Microsoft.Azure.Cosmos.ServiceInterop.dll" Pack="true" IsAssembly="true" PackagePath="runtimes\win-x64\native" />
+    <None Include="$(MSBuildThisFileDirectory)\Microsoft.Azure.Cosmos.Core.dll" Pack="true" IsAssembly="true" PackagePath="runtimes\any\lib\netstandard2.0" />
+    <None Include="$(MSBuildThisFileDirectory)\Microsoft.Azure.Cosmos.Serialization.HybridRow.dll" Pack="true" IsAssembly="true" PackagePath="runtimes\any\lib\netstandard2.0" />
+    <None Include="$(MSBuildThisFileDirectory)\Microsoft.Azure.Cosmos.Direct.dll" Pack="true" IsAssembly="true" PackagePath="lib\netstandard2.0" />
+    <None Include="$(MSBuildThisFileDirectory)\Microsoft.Azure.Cosmos.targets" Pack="true" PackagePath="build\netstandard2.0">
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: >Microsoft.Azure.Cosmos.targets [](start = 47, length = 30)

Do targets file cover service interop?

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 3 above)
Comment (invino4) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: +1

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj [hunk 7]
```diff
@@ -44,15 +44,38 @@
...
+    <PackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="4.5.1" />
+    <PackageReference Include="System.Threading.Tasks.Extensions" Version="4.5.1" />
+    <PackageReference Include="System.ValueTuple" Version="4.5.0" />
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(SignAssembly)' == 'true' ">
+    <!--CD will ensure below are present before signed release build-->
+    <None Include="$(MSBuildThisFileDirectory)\Cosmos.CRTCompat.dll" Pack="true" IsAssembly="true" PackagePath="runtimes\win-x64\native" />
+    <None Include="$(MSBuildThisFileDirectory)\Microsoft.Azure.Cosmos.ServiceInterop.dll" Pack="true" IsAssembly="true" PackagePath="runtimes\win-x64\native" />
+    <None Include="$(MSBuildThisFileDirectory)\Microsoft.Azure.Cosmos.Core.dll" Pack="true" IsAssembly="true" PackagePath="runtimes\any\lib\netstandard2.0" />
```
Comment (invino4) [inline Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: I'm not sure if these actually need to be in `runtimes\any\lib\netstandard2.0`.  I think `runtimes\lib\netstandard2.0` should be sufficient, no?

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 5 above)
Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: currently every package produced is signed

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 6 above)
Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: Yes, Microsoft.Azure.Cosmos.Serialization.HybridRow.dll and Microsoft.Azure.Cosmos.Core.dll

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 7 above)
Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: have not tried without "any" as "any" works and is the top level RID.

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 3 above)
Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: I'm just using it as a flag to determine package vs not package

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 4 above)
Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: I'm just using it as a flag to determine package vs not package

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 5 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: SignAssembly is only to conditional to decide which Direct version to use (MyGet or Nuget). Not for others.

**Comment by kirankumarkolli** [comment]
What's the PDB story?

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 7 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: Its not very clear from documentation but needs validation. https://stackoverflow.com/questions/52397501/nuget-references-to-assemblies-in-runtimes-folder-not-added/52471739#52471739 has some details. 

Form there it seems with out runtime identifier is used in FX not core.

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 3 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: Packaging and signing are two independent orthogonal ones.

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj [hunk 8]
```diff
@@ -44,15 +44,38 @@
...
   <ItemGroup Condition=" '$(SignAssembly)' != 'true' ">
     <PackageReference Include="Microsoft.Azure.Cosmos.Direct.MyGet" Version="[$(DirectVersion)]" />
+    <PackageReference Include="Microsoft.Azure.Cosmos.Serialization.HybridRow" Version="1.0.0-preview" />
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(SignAssembly)' == 'true' ">    
+    <Reference Include="Microsoft.Azure.Cosmos.Direct">
+      <HintPath>$(MSBuildThisFileDirectory)\Microsoft.Azure.Cosmos.Direct.dll</HintPath>
+    </Reference>
+    <!--Direct Dependencies-->
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: Isn't this what package manager are built for? It seems like Xamarin locks to a specific version for all of it's dependencies. 

https://www.nuget.org/packages/Xamarin.Android.Support.v4/
https://www.nuget.org/packages/Microsoft.CodeAnalysis.CSharp.Workspaces/
https://www.nuget.org/packages/Microsoft.NETCore.Runtime.CoreCLR-x64/1.0.1-beta-23225

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 7 above)
Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: "runtimes within sub-folders named {platform}-{architecture}\lib\{framework} or {platform}-{architecture}\native"

https://docs.microsoft.com/en-us/nuget/create-packages/supporting-multiple-target-frameworks
https://docs.microsoft.com/en-us/dotnet/core/rid-catalog

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 5 above)
Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: SignAssembly is used to sign the Microsoft.Azure.Cosmos assembly and as a flag for package vs not package. MyGet vs No MyGet, runtime HybridRow vs Package HybridRow etc.

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 3 above)
Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: It is only signed when packaged and only packaged when signed. The flag is used for both. No need for a separate flag.

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 1 above)
Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: @j82w It is the same except this is fragile and error prone. If a transitive dependency is introduced to Direct or Hybrid row you have to remember to add the dependency here. Also if you a new framework target is added to Direct or Hybrid row you have to remember to add it here and in the CI pipeline. Leadership has accepted the risk. Disagree and commit is the only option.

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 7 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: Only native are architecture specific (Win X64). 
Others are fully managed and should in lib hierarchy. 
Am I reading it right?

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 1 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: Eventually direct shouldn't exists (all code in a single repo) and also a single release cadence. Artificial separation now is reflecting process out (convoy :-) ). HR is interesting till it gets its own ship & hardened there are no other consumers hence its reasonable.

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 8 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: Related to above comments, please look at the comments there.

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.targets [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+      <Visible>False</Visible>
+    </ContentWithTargetPath>
+  </ItemGroup>
+  <ItemGroup>
+    <ContentWithTargetPath Include="$(MSBuildThisFileDirectory)..\..\runtimes\any\lib\netstandard2.0\Microsoft.Azure.Cosmos.Serialization.HybridRow.dll">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+      <TargetPath>Microsoft.Azure.Cosmos.Serialization.HybridRow.dll</TargetPath>
+      <Visible>False</Visible>
+    </ContentWithTargetPath>
+  </ItemGroup>
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.targets]: If they move to lib then its not necessary right?

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj (hunk 7 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: Also its odd that Direct is in lib where as HR&Core are in rutnime.

Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj]: That is why I used "any". If put in lib will will not be runtime, unless you create and empty assembly that you put into "ref" folder. As for Direct, I will put it into runtime for package generation.

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.targets (hunk 9 above)
Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.targets]: yes, but then will not be runtime and will be visible, unless you generate and empty assembly which is put into the "refs" folder.

**Pull request closed without merging**